rfcs/0013-configurable-and-composable-storage.md

```diff
@@ -0,0 +1,600 @@
+
+## Storage references
+
+[storage-references]: #storage-references
+
+Sometimes we want to be able to store in the storage a "reference" to another storage instance. Storage API and the compiler will provide support for such use cases. On the API level, the `StorageRef` type (defined in the [`storage.sw`](../files/0013-configurable-and-composable-storage/sway-libs/storage.sw)) will contain a type-safe "reference" to a storage element. This "reference" will internally just be the `StorageKey` of the referenced storage element.
+
+```Sway
+pub enum StorageRef<TStorage> where TStorage: Storage {
+    Null: (),
+    Ref: StorageKey,
+}
+```
+
+`StorageRef`s will be stored in [`StorageRefBox`es](../files/0013-configurable-and-composable-storage/sway-libs/storage/storage_ref_box.sw) and once retrieved from the storage it will be possible to dereference them to access the referenced storage.
+
+Every storage type automatically provides the `StorageRef` that references it, via the `Storage::as_ref` method.
+
+Thus, storage references provide a convenient, type-safe way to express referencing storage elements, but with a price that might be considerable in some cases. Namely, the `StorageRef` requires two storage slots to store a reference. This means that developers might consider other, less storage consuming, manual approaches to "reference" a storage entity. E.g., if referenced elements are stored in a `StorageVec`, "referencing" them by storing their vector indices as "references" might be less storage-costly. In this case, we are trading type-safety and clear built-in API for the performance.
+
+Storage reference will be supported in the `storage` declarations, by allowing the `storage` keyword to be used on the RHS of the _configure with_ operator:
+
+```Sway
+type StVecOfU64 = StorageVec<StorageBox<u64>>;
+type StPairOfStVecOfU64 = StoragePair<StVecOfU64, StVecOfU64>;
+
+storage {
+    vec_1: StVecOfU64 := [11, 22, 33],
+    vec_2: StVecOfU64 := [44, 55, 66],
+
+    pair_1: StPairOfStVecOfU64 := ([0, 0, 0], [1, 1, 1]),
+    pair_2: StPairOfStVecOfU64 := ([], []),
+
+    map_1: StorageMap<str[3], StVecOfU64> := [
+        ("abc", [1, 2, 3]),
+        ("def", [11, 22, 33]),
+        ("ghi", [111, 222, 333]),
+    ]
+
+    //--
+    // Storage references can be used in storage configurations.
+    // They can reference other storage elements, or their parts.
+    //--
+    vec_of_refs: StorageVec<StorageRefBox<StVecOfU64>> := [
+        StorageRef::Null,
+        storage.vec_1.as_ref(), // <<-- Note using the `storage` keyword here, as well as `as_ref`.
+        storage.pair_1.first().as_ref(),
+        storage.map_1.get("abc").as_ref(),
+        StorageRef::Ref(some_const_fn_that_returns_the_storage_key_of_the_referenced_storage()),
+    ],
+}
+```
+
+For more examples on using storage references, see the [`demo_contract_storage_references.sw`](../files/0013-configurable-and-composable-storage/contracts/demo_contract_storage_references.sw).
```

Comment:
I think `StorageRef` is an unnecessary complication.

I don't see what this adds over `Option<StorageKey>` which would already be supported. If anything it is worse because it reintroduces a potential implicit null.

Unless there's a compelling reason to have a special type, I say we should remove this and recommend storing `StorageKey`s and `Option<StorageKey>`s for references to other parts of the storage.

Reply:
Removing the enum and the `Null` and using `Option` instead I like a lot!

Actually, the first version of the `StorageRef` was a struct type-safely wrapping the `StorageKey` but then I ended up thinking how to support not having a reference and introduced `Null` motivated by the similar programming language concept. The result was putting the burden of checking for `Null` when dereferencing _always_, even in the situations where the reference will always be there. Using `Option<StorageRef>` where the `StorageRef` is a struct is definitely the right way!

To the question of having the dedicated `StorageRef`, or using the general `StorageKey` directly, the benefit of having the `struct StorageRef<TStorage>` is there, because of these two reasons:
1. Type-safety.
1. Having easy-to-use, zero-(additional)-cost abstraction that clearly communicates the intent and provides a standardized usage pattern.

Ad 1) `StorageRef<TStorage>` referrs to a storage of a particular type and that is type-checked by the compiler. It also clearly communicates that it referrs to a storage of a particular type. Using `StorageKey` would be more like having a raw pointer into the storage.

Ad 2) Using the naked `StorageKey` will require from developers to implement their own `deref` equivalents every time they use it for referencing. Those will be cumbersome and error-prone, because they will need to use the proper way to create the proper storage type etc. In the worse case, we will end up having that code in contracts, and in the best case, having it packed in various libraries. Having the `StorageRef` provided by the standard library gives a simple type that clearly communicates the intent and gives all convenience in a standardized and zero-cost way.